torch/testing/_asserts.py

```diff
@@ -313,7 +312,9 @@ def _check_values_close(
     Returns:
         (Optional[AssertionError]): If check did not pass.
     """
-
+    dtype = torch.float64 if actual.dtype.is_floating_point else torch.int64
```

Comment:
This would not work for complex tensors, but as long as we are checking the components individually, this is sufficient.

Reply:
How does assertEqual do this?

If we're comparing fp16 and fp64, for example, and the user doesn't specify rtol or atol, which rtol and atol are used? The one for fp16 or the one for fp64?

Reply:
This has nothing to do with the tolerances, since they are selected before we even hit that line. Plus, in `_check_values_close` the inputs are already equalized, so `actual` and `expected` have the same `dtype` here.

---

> How does assertEqual do this?

Internally, `self.assertEqual` calls `self._compareTensors` for the actual comparison. In there we hit 

https://github.com/pytorch/pytorch/blob/00896cb9ed931bb816cf1a5e5f13ddc6b229378b/torch/testing/_internal/common_utils.py#L1267-L1269

in case that we didn't specify the tolerances (recall that only specifying one is not allowed and will fail earlier). The default tolerances are selected by computing the maximum of both input dtypes:

https://github.com/pytorch/pytorch/blob/00896cb9ed931bb816cf1a5e5f13ddc6b229378b/torch/testing/_internal/common_utils.py#L1208-L1226

> If we're comparing fp16 and fp64, for example, and the user doesn't specify rtol or atol, which rtol and atol are used? The one for fp16 or the one for fp64?

The one for fp16. 

This is different what `assert_close` does now as observed in https://github.com/pytorch/pytorch/pull/58981#issuecomment-866654600, but I will send a PR that syncs the behavior.

Reply:
This is resolved in #60636 later in this stack.